examples/benchmark/main.rs

```diff
@@ -582,36 +585,21 @@ fn run_test(
             sleep(Duration::from_secs(5));
         }
 
-        let error_record = unsafe { queue.RetrieveErrorRecord() };
+        // TODO: inout or missing annotation?
```

Comment:
Do we need to open a ticket at the upstream DirectStorage project?

Reply:
@kennykerr can you help out with this? I recall some changes around `[Out]` parameters that can be optional, but I don't think this pointer is (it's not wrapped in `Option`).

This is the metadata in IlSpy:

```C#
unsafe void RetrieveErrorRecord([Out] DSTORAGE_ERROR_RECORD* record);
```

Seems relatively straightforward to translate to a return value, or can I override that in the metadata somehow?

Reply:
https://github.com/microsoft/DirectStorage/issues/38

Reply:
Or does it have something to do with the ABI? A function like https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12commandqueue-getdesc is implemented as I expect:

```rust
    pub unsafe fn GetDesc(&self) -> D3D12_COMMAND_QUEUE_DESC {
        let mut result__: D3D12_COMMAND_QUEUE_DESC = ::core::mem::zeroed();
        (::windows_core::Interface::vtable(self).GetDesc)(::windows_core::Interface::as_raw(self), &mut result__);
        result__
    }
````

But the definition in the metadata doesn't have an out parameter but a return type:

```rust
D3D12_COMMAND_QUEUE_DESC GetDesc();
```

Reply:
Please open a new issue if there's something you'd like me to look at, otherwise I will quickly forget this PR exists.

Reply:
Reported in https://github.com/microsoft/windows-rs/issues/2770.

Reply:
https://github.com/microsoft/windows-rs/issues/2770#issuecomment-1875651901 seems this is intended behaviour, if the struct is "large"?

Reply:
@hasenbanck this is fixed in the manual metadata now, by adding a `[RetVal]` annotation to the pointers: https://github.com/Tsukisoft/direct-storage-rs/commit/6a7f8c5d7e8b744bcf6f744ae661f8018528b6bf.

This does mean we'll have to generate the metadata (also to benefit from other improvements), and dodges confusing licensing for `winmd`. As pointed out in https://github.com/microsoft/DirectStorage/issues/8#issuecomment-1875760330 it is considered a header (that's what it was generated from after all), though the data type is binary type which I won't blatantly accept to fall under their MIT license for "code files".

---

One downside is that the custom metadata doesn't contain the `#define`s from `dstorageerr.h`. I don't know if that's a misconfiguration on my end, or if `WinmdGenerator` always skips `#define`s in favour of manually defining these enum constants in a `XXX.Manual.Cs` file. @damyanp suggestions?

Reply:
We could include the error's manually as it was the case before, until we find a better way.

Reply:
Sounds like a good idea, until we hear from Microsoft (@damyanp?) how they generate their metadata, and to incorporate our improvements.